web3.js/src/util/send-and-confirm-transaction.ts

```diff
@@ -33,17 +39,81 @@ export async function sendAndConfirmTransaction(
     sendOptions,
   );
 
-  const status = (
-    await connection.confirmTransaction(
-      signature,
-      options && options.commitment,
-    )
-  ).value;
+  const subscriptionCommitment =
+    options?.preflightCommitment || options?.commitment;
+  let subscriptionId: number | undefined;
+  let status: RpcResponseAndContext<SignatureResult> | null = null;
+  let message: string | null = null;
+
+  const confirmTx = new Promise((resolve, reject) => {
+    try {
+      subscriptionId = connection.onSignature(
+        signature,
+        (result: SignatureResult, context: Context) => {
+          subscriptionId = undefined;
+          status = {
+            context,
+            value: result,
+          };
+          message = 'confirmed';
+          resolve(message);
+        },
+        subscriptionCommitment,
+      );
+    } catch (err) {
+      reject(err);
+    }
+  });
 
-  if (status.err) {
-    throw new Error(
-      `Transaction ${signature} failed (${JSON.stringify(status)})`,
+  const checkBlockHeight = async () => {
+    const blockHeight = await connection.getBlockHeight(
+      options && options.commitment,
     );
+
+    return blockHeight;
+  };
+
+  const expireTx = new Promise((resolve, reject) => {
+    (async () => {
+      try {
+        let currentBlockHeight = await checkBlockHeight();
+        while (currentBlockHeight <= transaction.lastValidBlockHeight!) {
+          if (transaction.lastValidBlockHeight! - currentBlockHeight > 200) {
+            await sleep(2000);
+          } else if (transaction.lastValidBlockHeight! - currentBlockHeight > 100) {
+            await sleep(1000);
+          } else {
+            await sleep(500);
+          }
+          currentBlockHeight = await checkBlockHeight();
+        }
+        console.log('lastValidBlockHeight exceeded!');
+        message = 'expired';
+        resolve(message);
+      } catch (error) {
+        reject(error);
+      }
+    })();
+  });
+
+  try {
+    const outcome = await Promise.race([confirmTx, expireTx]);
+    switch (outcome) {
+      case 'confirmed':
+        console.log('Transaction has been confirmed');
+        break;
+      case 'expired':
+        console.log('Transaction has expired');
+        break;
+    }
+  } finally {
+    if (subscriptionId) {
+      connection.removeSignatureListener(subscriptionId);
+    }
+  }
+
+  if (status === null) {
+    throw new Error('Transaction has expired.');
```

Comment:
* Include the signature in this error message.
* I'd prefer if we started to throw first-class exceptions. That way people can catch them and do `instanceof` checks on them.

```
class TransactionExpiredBlockheightExceededError extends Error {
  constructor(signature) {
    super(`Signature: ${signature}`);
    this.signature = signature;
  }
}
Object.defineProperty(TransactionExpiredBlockheightExceededError.prototype, 'name', {
  value: 'TransactionExpiredBlockheightExceededError',
});

if (...) {
  throw new TransactionExpiredBlockheightExceededError(signature);
}
```

Reply:
Okay, I'll work on this. Thanks!